packages/expo/cli/start/server/AsyncNgrok.ts

```diff
@@ -0,0 +1,186 @@
+      shouldPrompt: false,
+      autoInstall: false,
+    });
+
+    let timer: NodeJS.Timeout | null = null;
+    try {
+      // TODO(Bacon): Consider dropping the timeout functionality:
+      // https://github.com/expo/expo/pull/16556#discussion_r822307373
+      const results = await Promise.race([
+        // Returns false to try again.
+        this.connectToNgrokInternalAsync(instance, attempts),
+
+        // A timeout that correctly surfaces the error to the correct lexical scope.
+        new Promise<false>((_, reject) => {
+          timer = setTimeout(() => {
+            reject(new CommandError('NGROK_TIMEOUT', 'ngrok tunnel took too long to connect.'));
+          }, options.timeout ?? TUNNEL_TIMEOUT);
+        }),
+      ]);
+      if (typeof results === 'string') {
+        return results;
+      }
+    } finally {
+      clearTimeout(timer);
+    }
+    // Wait 100ms and then try again
+    await delayAsync(100);
+
+    return this._connectToNgrokAsync(options, attempts + 1);
+  }
+
+  private async connectToNgrokInternalAsync(
+    instance: NgrokInstance,
+    attempts: number = 0
+  ): Promise<string | false> {
+    try {
+      // Global config path.
+      const configPath = path.join(UserSettings.getDirectory(), 'ngrok.yml');
+      Log.debug('[ngrok] Global config path:', configPath);
+      const hostname = await this._getProjectHostnameAsync();
+      Log.debug('[ngrok] Hostname:', hostname);
+
+      const url = await instance.connect({
+        authtoken: NGROK_CONFIG.authToken,
+        proto: 'http',
+        hostname,
+        configPath,
+        onStatusChange(status) {
+          if (status === 'closed') {
+            Log.error(
+              'We noticed your tunnel is having issues. ' +
+                'This may be due to intermittent problems with ngrok. ' +
+                'If you have trouble connecting to your app, try to restart the project, ' +
+                'or switch the host to `lan`.'
+            );
+          } else if (status === 'connected') {
+            Log.log('Tunnel connected.');
+          }
+        },
+        port: this.port,
+      });
+      return url;
+    } catch (error: any) {
+      // Attempt to connect 3 times
+      if (attempts >= 2) {
+        throw new CommandError('NGROK_CONNECT', error.toString());
+      }
+
+      // Attempt to fix the issue
+      if (error?.error_code === 103) {
+        // Change randomness to avoid conflict if killing ngrok doesn't help
+        await this._resetProjectRandomnessAsync();
+      }
+      return false;
+    }
+  }
+
+  private async getProjectRandomnessAsync() {
+    const { urlRandomness: randomness } = await ProjectSettings.readAsync(this.projectRoot);
+    if (randomness) {
+      return randomness;
+    }
+    return await this._resetProjectRandomnessAsync();
+  }
+
+  async _resetProjectRandomnessAsync() {
+    const randomness = crypto.randomBytes(5).toString('base64url');
```

Comment:
I believe base64url is supported only in Node 16+

Reply:
Introduced in v14.18.0 which would be LTS by the time we release the CLI. https://nodejs.org/api/buffer.html